### PR TITLE
[ESI][Runtime] Generate C++ header files for constants

### PR DESCRIPTION
**Pull request by teqdruid**
Start of static C++ header file generation. Just integer constants for now. Can work off of any manifest and will even connect to a live accelerator and read the manifest from there.

Generates a `types.h` file and one per module listed in the manifest. Inside the module header file, generates one class per module and adds the constants to that class. Puts _everything_ in a namespace specified by the user as `system_name`.

### CODE REVIEW

File: lib/Dialect/ESI/runtime/python/esiaccel/cppgen.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+
+import argparse
+from pathlib import Path
+import textwrap
+import sys
+
+_thisdir = Path(__file__).absolute().resolve().parent
+
+
+class Generator:
```
Comment (mortbopet) [inline lib/Dialect/ESI/runtime/python/esiaccel/cppgen.py:21]: move to other file?

File: lib/Dialect/ESI/runtime/python/esiaccel/cppgen.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    """Get the C++ code for a constant in a module."""
+    const_strs: List[str] = [
+        f"static constexpr {self.get_type_str(const.type)} "
+        f"{name} = 0x{const.value:x};"
+        for name, const in module_info.constants.items()
+    ]
+    return "\n".join(const_strs)
+
+  def write_modules(self, output_dir: Path, system_name: str):
+    """Write the C++ one header for each module in the manifest."""
```
Comment (mortbopet) [inline lib/Dialect/ESI/runtime/python/esiaccel/cppgen.py:64]: Rephrase

File: lib/Dialect/ESI/runtime/python/esiaccel/cppgen.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+    for module_info in self.manifest.module_infos:
+      s = f"""
+      /// Generated header for {system_name} module {module_info.name}.
+      #pragma once
+      #include "types.h"
```
Comment (mortbopet) [inline lib/Dialect/ESI/runtime/python/esiaccel/cppgen.py:70]: Way too generic, and alias-prone. `ESITypes.h` at a minimum. Ideally, there'd be some include path here (`#include "ESI/types.h"`).

File: lib/Dialect/ESI/runtime/python/esiaccel/cppgen.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+      """))
+
+  argparser.add_argument("--file",
+                         type=str,
+                         default=None,
+                         help="Path to the manifest file.")
+  argparser.add_argument(
+      "--platform",
+      type=str,
+      help="Name of backend for live accelerator connection.")
```
Comment (mortbopet) [inline lib/Dialect/ESI/runtime/python/esiaccel/cppgen.py:145]: platform or backend? Seems like it's referring to backends, but maybe im misunderstanding the scope of backends/(platforms?).

**Review by mortbopet — COMMENTED** [review]
Missing a test/check on the generated .h file. Would also make this PR a bit easier to review if there was something that one could inspect to see the output of the generator.

File: lib/Dialect/ESI/runtime/python/esiaccel/cppgen.py (hunk 3 above)
Comment (teqdruid) [reply lib/Dialect/ESI/runtime/python/esiaccel/cppgen.py:70]: The idea is that users could choose the name of the folder in which to output the files. (e.g. `ESI/` or `<system-specific-name>/`) so the output file would be something like `ESI/types.h`.

File: lib/Dialect/ESI/runtime/python/esiaccel/cppgen.py (hunk 4 above)
Comment (teqdruid) [reply lib/Dialect/ESI/runtime/python/esiaccel/cppgen.py:145]: Backend ~= platform. Technically, a backend is a particular runtime implementation of an FPGA platform. I do expect a 1:1 mapping in most cases. I tend to refer to platform for anything user facing.

**Review by mortbopet — APPROVED** [review]

**Pull request merged**